### ISSUE TITLE
Multiple scaleobject on the same namespace for azure pipelines

### ISSUE THREAD

**Issue by shzshi** [kedacore/keda#4014]
### Report

When I created two scaledobjects for different scaleTargetRef with two different deployment on the same namespace , the HPA scales for both the deployments agents instead of scaling the specific scaleTargetRef 

### Expected Behavior

The scaling of the agent should happen just to specific scaleTargetRef mentioned within the scaleobject

### Actual Behavior

When I created two scaledobjects for different scaleTargetRef with two different deployment on the same namespace , the HPA scales for both the deployments agents instead of scaling the specific scaleTargetRef 

### Steps to Reproduce the Problem

1. create two scaledObjectsand TriggerAuthentications within same namespace for two deployment of ado agent within the agent same pool
2. Test with sample pipeline pointing to specific agent through demand eg : API_VERSION equals 17 
3. the agents for both the ado agents will scale although scaledObject , HPA, TriggerAuthetication and ScaledObjectRef are separate.   

### Logs from KEDA operator

```
example
```

### KEDA Version

2.8.1

### Kubernetes Version

1.25

### Platform

Microsoft Azure

**Comment by shzshi** [comment]
Hi , 

please find the details below. 

scaleoject 1 
--------------
apiVersion: keda.sh/v2
kind: TriggerAuthentication
metadata:
  name: pipeline-trigger-auth
  namespace: az-devops
spec:
  secretTargetRef:
    - parameter: personalAccessToken
      name: azdevopsecret 
      key: AZP_TOKEN # personalAccessToken
---
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: azure-pipelines-scaledobject
  namespace: az-devops
spec:
  scaleTargetRef:
    name: "azdevops-deployment-keda"
  minReplicaCount: 1
  maxReplicaCount: 5 
  cooldownPeriod: 3
  triggers:
  - type: azure-pipelines
    metadata:
      poolID: "1"
      poolName: "MY-POOL_NAME"
      organizationURLFromEnv: "AZP_URL"
      targetPipelinesQueueLength: "1"
    authenticationRef:
     name: pipeline-trigger-auth
     
------------------------------------------------------
 scaleobject2
------------------------------------------------------
apiVersion: keda.sh/v2
kind: TriggerAuthentication
metadata:
  name: pipeline-trigger-auth-node
  namespace: az-devops
spec:
  secretTargetRef:
    - parameter: personalAccessToken
      name: azdevopsecret # pipeline-auth
      key: AZP_TOKEN # personalAccessToken
---
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: azure-pipelines-scaledobject-node
  namespace: az-devops
spec:
  scaleTargetRef:
    name: azdevops-deployment-node
  minReplicaCount: 1
  maxReplicaCount: 5 
  cooldownPeriod: 3
  triggers:
  - type: azure-pipelines
    metadata:
      poolID: "1"
      poolName: "MY-POOL_NAME"
      organizationURLFromEnv: "AZP_URL"
      targetPipelinesQueueLength: "1"
      demands: "NODE_VERSION"
    authenticationRef:
     name: pipeline-trigger-auth-node

**Comment by tomkerkhove** [comment]
That might be a bug, what do you think @JorTurFer ?

**Comment by JorTurFer** [comment]
IDK, it's weird tbh, I'll try to reproduce it during the week

**Comment by shzshi** [comment]
Hi Team, 

Any update on this ? 

Cheers,
Shashi

**Comment by JorTurFer** [comment]
I haven't tested this yet, pre-Christmas usually is crazy :(

**Comment by JorTurFer** [comment]
I still have to test this :(

**Comment by shzshi** [comment]
any update on this ?

**Comment by JorTurFer** [comment]
I'd like to test this during this week before release, sorry :(

**Comment by shzshi** [comment]
thank you , counting on you :)

**Comment by JorTurFer** [comment]
Hi, I haven't been able to reproduce it but I think that I know why:
You only have set `demands` in one of them. This means that the ScaledObject without any demands will scale based on the queue length without any other filter. You have 2 ScaledObjects, one scaling based on a specific demand and another based on any demand. 
They don't exclude themselves, the scaler checks for not in progress jobs, but you could have both scaling at the same time. 
I guess that it's something misconfigured, but I can't reproduce it so I think it's not related with KEDA itself

**Comment by shzshi** [comment]
even if I remove demand from scaledBoject , I still see scaling for both the deployments at same time , if there is load on just one deployment , I will share the logs tomorrow.

**Comment by JorTurFer** [comment]
Yes, because you are scaling based on the same queue and even though we discard in-progress jobs, agents need some time for registering the job as completed. What does it happen if you set demands in both and schedule a job that only one can address?

**Comment by shzshi** [comment]
i didn't try demand for both scaled object, but did try parent with pod name and it works fine , but as pod name is not constant , so that solution won't work.

**Comment by JorTurFer** [comment]
@Eldarrin , do you have any idea about what could be happening?

**Comment by Eldarrin** [comment]
It’s as you say on demands. Each scaler is its own object and they fire at the same time so effectively pull the same list of job requests. If using demands, they must be exclusive; otherwise 1 fires if the demand is satisfied and 1 fires because it’s always satisfied.
When using demands I would suggest on agent A you set an env variable like “NODEJS16=true” and on agent B you set an env variable like “NODESJ18=true” then in your ado pipeline you would set:

pool:
  name: POOLNAME
  demands: nodejs16

In your ScaledObject meta.demands you put nodejs16 for one and nodejs18 for the other.

If using the parent method, rather than the pod name, set the env variable AZP_AGENT_NAME in your reference container.

**Comment by shzshi** [comment]
if I use AZP_AGENT_NAME  then ado will create agent with that name and scaling breaks , because agent will have same name

**Comment by shzshi** [comment]
yes I have set that as envronment variable on the azure devops kubernetes deployment , something like below and then referencing it as parent in the scaledObject, but when make this changes and deploy again then new pod agent is getting created , but as pod name or agent is name is not dynamic and which is why scaling does not work. 

------------------------
agent-deployment.yaml
------------------------
- name: AZP_AGENT_NAME
            valueFrom:
              secretKeyRef:
                name: azdevops
                key: AZP_AGENT_NAME
                
-------------------
ScaledObject
-------------------
metadata: 
   parent: "AZP_AGENT_NAME"

**Comment by Eldarrin** [comment]
https://github.com/Eldarrin/keda-azdo-example

Here is an example of how I use it in production. Mine has a lot more around it of course, but the principal is the same.

**Comment by JorTurFer** [comment]
Hi @shzshi ,
Is the example enough?

**Comment by shzshi** [comment]
Hi JorTurFer, 

Apologies for delayed , I applied same thing but it didn't work :(

**Comment by JorTurFer** [comment]
I have tested again and I can't replicate it :( 
Could you share sample manifests reproducing your scenario? I mean, SOs, deploys, etc. I'd need also the agent capabilities to reproduce the scenario
